### PR TITLE
Lints on AST, remove ExprParen

### PR DESCRIPTION
**Pull request by nrc**
r? @nikomatsakis, f? @Manishearth

### CODE REVIEW

**Comment by Manishearth** [comment]
I'd personally prefer if there were different kinds of lints entirely (not one kind of lint with different available contexts -- this means the lint will be run multiple times). So, basically, the LintStore would contain both EarlyLintPasses and LateLintPasses. It also keeps the lint traits easier to understand. If we want to have "MIR lints" in the future, the design is easily extended.

But I'm not insistent on this, the current model is fine too.

File: src/librustc/lint/mod.rs [hunk 1]
```diff
@@ -126,46 +128,92 @@ pub trait LintPass {
...
+    fn check_struct_field(&mut self, _: &LateContext, _: &hir::StructField) { }
+    fn check_variant(&mut self, _: &LateContext, _: &hir::Variant, _: &hir::Generics) { }
+    fn check_variant_post(&mut self, _: &LateContext, _: &hir::Variant, _: &hir::Generics) { }
+    fn check_opt_lifetime_ref(&mut self, _: &LateContext, _: Span, _: &Option<hir::Lifetime>) { }
+    fn check_lifetime_ref(&mut self, _: &LateContext, _: &hir::Lifetime) { }
+    fn check_lifetime_def(&mut self, _: &LateContext, _: &hir::LifetimeDef) { }
+    fn check_explicit_self(&mut self, _: &LateContext, _: &hir::ExplicitSelf) { }
+    // Note that you shouldn't implement both check_mac and check_ast_mac,
+    // because then your lint will be called twice. Prefer check_ast_mac.
+    fn check_mac(&mut self, _: &LateContext, _: &ast::Mac) { }
```
Comment (Manishearth) [inline src/librustc/lint/mod.rs]: Why would we have a macro in the HIR tree? Isn't HIR post-expansion?

Comment (nikomatsakis) [inline src/librustc/lint/mod.rs]: @Manishearth 

> Why would we have a macro in the HIR tree? Isn't HIR post-expansion?

I believe these are exported macro definitions, not macro uses.

**Comment by nrc** [comment]
My first idea for this was having two kinds of lints (early and late), but I decided against basically for backwards compatibility and simplicity. With the changes as they stand, lint authors only need to import a trait. I thought having to use different pass names, etc. would be too much of a breaking change. However, I guess if we are going to break lints even a little, then breaking them a lot is not much worse. I also thought it was easier having one big lint, rather than lots of little ones, but I'm not really sure about that.

If the backcompat issue is not too big a deal, then I can redo this with different lint passes pretty easily.

File: src/librustc/lint/mod.rs (hunk 1 above)
Comment (Manishearth) [inline src/librustc/lint/mod.rs]: @nikomatsakis http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.Mac_.html

Docs say otherwise

**Comment by Manishearth** [comment]
I don't think it would be a major backcompat issue; given that it's just one trait that can be sed'ed (most lints work at the HIR stage). Only AST lints will have some backcompat issues (which is a tiny bit more sed)

**Comment by petrochenkov** [comment]
I'm waiting for this to land and then will try to finish https://github.com/rust-lang/rust/issues/6993 by replacing `Ident`s with `Name`s in HIR and below.

**Comment by nikomatsakis** [comment]
I defer to @Manishearth (aka Mr Lint) when it comes to the backwards compat pain. Basically seems like we ought to redo this PR as described?

**Comment by Manishearth** [comment]
I'm okay with the status quo, but I'd prefer the multitrait approach, yeah. I don't think it would be a major backcompat issue. Most lints out there are conceptually HIR lints, and we'd just need to replace some trait imports.

The original HIR changes (which had a lot of fiddly, non-sed-able import changes) were larger backcompat issues than this could be.

**Comment by nrc** [comment]
Now with multiple traits

**Comment by Manishearth** [comment]
LGTM

**Comment by Manishearth** [comment]
```
../src/librustc/middle/expr_use_visitor.rs:283:22: 287:10 error: cannot infer an appropriate lifetime for lifetime parameter `'tcx` due to conflicting requirements
../src/librustc/middle/expr_use_visitor.rs:283         let result = ExprUseVisitor {
../src/librustc/middle/expr_use_visitor.rs:284             typer: typer,
../src/librustc/middle/expr_use_visitor.rs:285             mc: mc::MemCategorizationContext::new(typer),
../src/librustc/middle/expr_use_visitor.rs:286             delegate: delegate,
../src/librustc/middle/expr_use_visitor.rs:287         };
../src/librustc/middle/expr_use_visitor.rs:279:5: 290:6 help: consider using an explicit lifetime parameter as shown: fn new(delegate: &'d mut Delegate<'tcx>, typer: &'t infer::InferCtxt<'a, 't>)
 -> ExprUseVisitor<'d, 't, 'a, 't>
../src/librustc/middle/expr_use_visitor.rs:279     pub fn new(delegate: &'d mut Delegate<'tcx>,
../src/librustc/middle/expr_use_visitor.rs:280                typer: &'t infer::InferCtxt<'a, 'tcx>)
../src/librustc/middle/expr_use_visitor.rs:281                -> ExprUseVisitor<'d,'t,'a,'tcx>
../src/librustc/middle/expr_use_visitor.rs:282     {
../src/librustc/middle/expr_use_visitor.rs:283         let result = ExprUseVisitor {
../src/librustc/middle/expr_use_visitor.rs:284             typer: typer,
                                               ...
error: aborting due to previous error
```

**Pull request merged**